src/login.rs

```diff
@@ -182,20 +211,69 @@ impl Login {
 
         let self_ = imp::Login::from_instance(self);
         let visible_page = self_.content.visible_child_name().unwrap();
-        if visible_page == "phone-number-page" {
-            let encryption_key = self_.custom_encryption_key_entry.text().to_string();
-            if !encryption_key.is_empty() {
-                self.change_encryption_key();
+
+        match visible_page.as_str() {
+            "phone-number-page" => {
+                let encryption_key = self_.custom_encryption_key_entry.text().to_string();
+                if !encryption_key.is_empty() {
+                    self.change_encryption_key();
+                }
+
+                self.send_phone_number();
+            }
+            "code-page" => self.send_code(),
+            "registration-page" => {
+                if !self_.show_tos_popup.get() || self_.tos_accepted_by_user.get() {
+                    // Just proceed if the user either doesn't need to accept the ToS or has
+                    // already accepted it
+                    self.send_registration()
+                } else {
+                    // Force the ToS dialog for the user before he can proceed
+                    self.show_tos_dialog(true);
+                }
             }
+            "password-page" => self.send_password(),
+            "encryption-key-page" => self.send_encryption_key(false),
+            other => unreachable!("no page named '{}'", other),
+        }
+    }
+
+    fn show_tos_dialog(&self, user_wants_to_proceed: bool) {
+        let self_ = imp::Login::from_instance(self);
 
-            self.send_phone_number();
-        } else if visible_page == "code-page" {
-            self.send_code();
-        } else if visible_page == "password-page" {
-            self.send_password();
-        } else if visible_page == "encryption-key-page" {
-            self.send_encryption_key(false);
+        let builder = gtk::MessageDialog::builder()
+            .use_markup(true)
+            .secondary_text(&*self_.tos_text.borrow())
+            .modal(true)
+            .transient_for(self.root().unwrap().downcast_ref::<gtk::Window>().unwrap());
+
+        let dialog = if self_.show_tos_popup.get() {
```

Comment:
```suggestion
        let dialog = if user_needs_to_accept {
```
Otherwise the user would get the Yes/No dialog even when he/she wants to only see the tos pressing the label at the bottom.

Reply:
Yeah, my intention was that the user could accept the ToS beforehand. So that she/he will not be bothered by the ToS dialog again. :smiley: